src/math/interpolator.hh

```diff
@@ -0,0 +1,1168 @@
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3, or (at
+ * your option) any later version.
+ *
+ * rascal is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this software; see the file LICENSE. If not, write to the
+ * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+ * Boston, MA 02111-1307, USA.
+ */
+
+#ifndef SRC_MATH_INTERPOLATOR_HH_
+#define SRC_MATH_INTERPOLATOR_HH_
+
+#include <functional>
+#include <forward_list>
+#include <iostream>
+#include <limits>
+#include <cassert>
+#include "math_utils.hh"
+
+namespace rascal {
+  namespace math {
+    using Vector_Ref = typename Eigen::Ref<const Vector_t>;
+    using Matrix_Ref = typename Eigen::Ref<const Matrix_t>;
+
+    enum class GridType_t { Uniform };
+    enum class RefinementMethod_t { Exponential, Linear, Adaptive };
+
+    template <GridType_t Type, RefinementMethod_t Method>
+    struct GridRational {};
+
+    template <>
+    struct GridRational<GridType_t::Uniform, RefinementMethod_t::Adaptive> {
+      GridRational<GridType_t::Uniform, RefinementMethod_t::Adaptive>()
+          : grid_meshes{}, grid_meshes_error{}, grid_size{0}, max_it{} {}
+
+      constexpr static GridType_t GridType{GridType_t::Uniform};
+      constexpr static RefinementMethod_t RefinementMethod{
+          RefinementMethod_t::Adaptive};
+      // These are possible improvements for this grid rational:
+      // - for this grid rational it makes sense to save the test grid,
+      // because it is one step ahead and we can add more points
+      // - make hyperparameter k highest error
+      // - make grid rational init compute, increase_finness() and then remove
+      // x1,x2,grid_fineness
+      Vector_t compute_grid(double x1, double x2, int grid_fineness) {
+        if (grid_fineness == 0) {
+          this->grid_meshes = {x1, x2};
+          this->grid_size = 2;
+          return this->grid_from_meshes();
+        }
+        this->refine();
+        return this->grid_from_meshes();
+      }
+
+      void refine() {
+        auto next_it{this->max_it};
+        next_it++;
+        double cur{*this->max_it};
+        double mid_point{cur + (*next_it - cur) / 2};
+        this->grid_meshes.emplace_after(this->max_it, mid_point);
+        this->grid_size++;
+      }
+
+      Vector_t grid_from_meshes() {
+        Vector_t grid = Vector_t::Zero(this->grid_size);
+        int i{0};
+        for (auto it = this->grid_meshes.begin(); it != this->grid_meshes.end();
+             ++it) {
+          grid(i) = (*it);
+          i++;
+        }
+        return grid;
+      }
+
+      Vector_t compute_test_grid(double, double, int) {
```

Comment:
I would prefer commented variable names here, e.g. `double /*x1*/`.
At least in other places we do have used this.